src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1436 @@
+    /* No further action is needed since the currently 
+    available methods evaluate the RHS at the end of each time step*/
+
+    break;
+
+  case ARK_FULLRHS_OTHER:
+
+    /* call f */
+    retval = step_mem->fe(t, y, f, ark_mem->user_data);
+    step_mem->nfe++;
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                      MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
+
+    break;
+
+  default:
+    /* return with RHS failure if unknown mode is passed */
+    arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                    "Unknown full RHS mode");
+    return (ARK_RHSFUNC_FAIL);
+  }
+
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_TakeStepRKC:
+
+  This routine serves the primary purpose of the LSRKStepRKC module:
+  it performs a single RKC step (with embedding, if possible).
+
+  The output variable dsmPtr should contain estimate of the
+  weighted local error if an embedding is present; otherwise it
+  should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step.  As this routine
+  involves no algebraic solve, it is set to 0 (success).
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int lsrkStep_TakeStepRKC(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+{
+  int retval;
+  sunrealtype* cvals;
+  sunrealtype w0, w1, temp1, temp2, arg, bjm1, bjm2, mus, thjm1, thjm2, zjm1,
+    zjm2, dzjm1, dzjm2, d2zjm1, d2zjm2, zj, dzj, d2zj, bj, ajm1, mu, nu, thj;
+  sunrealtype onep54 = 1.54, c13 = 13.0, p8 = 0.8, p4 = 0.4;
+  N_Vector* Xvecs;
+  ARKodeLSRKStepMem step_mem;
+
+  /* initialize algebraic solver convergence flag to success,
+     temporal error estimate to zero */
+  *nflagPtr = ARK_SUCCESS;
+  *dsmPtr   = ZERO;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  cvals = step_mem->cvals;
+  Xvecs = step_mem->Xvecs;
+
+  /* Compute Dominated Eigenvalue and update stats */
+  if ((step_mem->newdomeig))
+  {
+    retval = lsrkStep_ComputeNewDomEig(ark_mem, step_mem);
+    if (retval != ARK_SUCCESS) { return (retval); }
+    step_mem->ndomeigupdates++;
+  }
+
+  /* determine the number of required stages */
+  for (int ss = 1; ss < step_mem->stagemaxlimit; ss++)
+  {
+    if (SUNSQR(ss) >= (onep54 * SUNRabs(ark_mem->h) * step_mem->sprad))
+    {
+      step_mem->reqstages = SUNMAX(ss, 2);
+      break;
+    }
+  }
+  step_mem->stagemax = SUNMAX(step_mem->reqstages, step_mem->stagemax);
```

Comment:
For both RKC and RKL (and later RKG), I think that we should do something to handle the case where `stagemaxlimit` is too small to contain h*lambda.  As it stands, you'll just use the maximum allowed number of stages *even if you expect this to be unstable*.  I think it might be preferable to just return from the TakeStep function with a recoverable failure, so that ARKODE can reduce h and then try again, without wasting all of the computational effort required to compute the unstable step.

Reply:
I have an update where I try to resolve this issue within the takestep. Could you please check it and let me know your comments?

Reply:
It looks like you only did this in `lsrkStep_TakeStepRKL` (but not `lsrkStep_TakeStepRKC`.  Whatever we settle on there should be replicated in both locations.

Unfortunately, we do not allow a stepper to adjust the step size directly (the step size itself is "owned" by ARKODE).  Here's what I was thinking:
#. You do your normal loop to set `reqstages`
#. If we reach the final iteration of that loop and we have not set the value, then you determine the maximum value of `h` that *would* be possible based on `stagemaxlimit` and `sprad`.  Let's call this value `maxh`.  
#. We then store that recommended step size (e.g., using `ark_mem->eta = maxh / ark_mem->h`), and return some *new* positive return flag.

From looking at `arkode.c`, lines 871-951, I believe that this alone would cause ARKODE to reattempt the step with the recommended step size `hmax`, without causing any other issues.